### PR TITLE
feat(TFD-8345): add fieldname

### PR DESCRIPTION
**Pull request by pbailly**
Associate a field name to indexed record.
This allows us to infer field name on field selector.

### CODE REVIEW

File: project/Build.scala [hunk 1]
```diff
@@ -19,7 +19,7 @@ object Build extends sbt.Build {
 
   lazy val basicSettings = Seq(
     organization := "com.wandoulabs.avro",
-    version := "0.1.3-talend-SNAPSHOT",
+    version := "0.1.4-talend-SNAPSHOT",
```
Comment (pbailly) [inline project/Build.scala:22]: TODO: Make a release.

File: src/main/scala/wandou/avpath/Evaluator.scala [hunk 2]
```diff
@@ -376,26 +403,27 @@ object Evaluator {
...
             }
-          case Ctx(arr: java.util.Collection[_], schema, topLevelField, _) =>
+          case Ctx(arr: java.util.Collection[_], name, schema, topLevelField, path, _) =>
             getElementType(schema) foreach { elemType =>
               val values = arr.iterator
               var j = 0
               while (values.hasNext) {
                 val value = values.next
-                res ::= Ctx(value, elemType, topLevelField, Some(TargetArray(arr, j, schema)))
+                // not sure
```
Comment (pbailly) [inline src/main/scala/wandou/avpath/Evaluator.scala]: We are on a list, we name each sub element like the parent.
It should not be troublesome, but it's not right.

File: src/main/scala/wandou/avpath/Evaluator.scala [hunk 3]
```diff
@@ -424,19 +452,20 @@ object Evaluator {
...
 
-      case Ctx(arr: java.util.Collection[_], schema, topLevelField, _) =>
+      case Ctx(arr: java.util.Collection[_], name, schema, topLevelField, path, _) =>
         getElementType(schema) foreach { elemType =>
           val values = arr.iterator
           var i = 0
           while (values.hasNext) {
             val value = values.next
-            val elemCtx = Ctx(value, elemType, topLevelField, Some(TargetArray(arr, i, schema)))
+            // not sure
```
Comment (pbailly) [inline src/main/scala/wandou/avpath/Evaluator.scala]: We are on a list, we name each sub element like the parent.
It should not be troublesome, but it's not right.

File: src/main/scala/wandou/avpath/Evaluator.scala [hunk 4]
```diff
@@ -445,12 +474,13 @@ object Evaluator {
...
         }
 
-      case Ctx(map: java.util.Map[String, _] @unchecked, schema, topLevelField, _) =>
+      case Ctx(map: java.util.Map[String, _] @unchecked, name, schema, topLevelField, path, _) =>
         getValueType(schema) foreach { elemType =>
           val entries = map.entrySet.iterator
           while (entries.hasNext) {
             val entry = entries.next
-            val elemCtx = Ctx(entry.getValue, elemType, topLevelField, Some(TargetMap(map, entry.getKey, schema)))
+            // TODO not sure
```
Comment (pbailly) [inline src/main/scala/wandou/avpath/Evaluator.scala]: We are on a Map, we name each sub element like the parent.
It should not be troublesome, but it's not right.

File: src/main/scala/wandou/avpath/Evaluator.scala [hunk 5]
```diff
@@ -479,7 +509,8 @@ object Evaluator {
               var i = 0
               while (values.hasNext) {
                 val value = values.next
-                elems(i) = Ctx(value, elemType, topLevelField, Some(TargetArray(arr, i, schema)))
+                // TODO not sure
```
Comment (pbailly) [inline src/main/scala/wandou/avpath/Evaluator.scala]: We are on a list, we name each sub element like the parent.
It should not be troublesome, but it's not right.

File: src/main/scala/wandou/avpath/Evaluator.scala [hunk 6]
```diff
@@ -16,70 +16,97 @@ import wandou.avpath.Parser.Syntax
 import wandou.avpath.Parser.UnaryExprSyntax
 import wandou.avpath.Parser.PosSyntax
 import java.nio.ByteBuffer
+import java.util.regex.Pattern
```
Comment (RyanSkraba) [inline src/main/scala/wandou/avpath/Evaluator.scala]: Is there any reason not to use the `scala.util.matching.Regex` idioms instead?  (See below)

**Review by RyanSkraba — CHANGES_REQUESTED** [review]

File: project/Build.scala (hunk 1 above)
Comment (RyanSkraba) [reply project/Build.scala:22]: Just for info, current version being sent with processing-beam is `0.1.3-talend`, which is just a re-labeled `0.1.3-talend-SNAPSHOT`

File: src/main/scala/wandou/avpath/Evaluator.scala [hunk 7]
```diff
@@ -16,70 +16,97 @@ import wandou.avpath.Parser.Syntax
...
     case object Insert extends Op
+
     case object InsertAll extends Op
+
   }
+
   private sealed trait Op
 
-  final case class Ctx(value: Any, schema: Schema, topLevelField: Schema.Field, target: Option[Target] = None)
+  val pattern: Pattern = Pattern.compile("\\[(\\*|:-?\\d+|-?\\d+:|-?\\d+:-?\\d+)\\]")
```
Comment (RyanSkraba) [inline src/main/scala/wandou/avpath/Evaluator.scala]: val date = """(\d\d\d\d)-(\d\d)-(\d\d)""".r
```suggestion
  import scala.util.matching.Regex
  // Matches any [*], [:n], or [n:m] range selectors.
  val pattern: Regex = """\[(\*|:-?\d+|-?\d+:|-?\d+:-?\d+)\\]""".re
```

It looks like the scala regexs are pretty clean for use in scala.

File: src/main/scala/wandou/avpath/Evaluator.scala [hunk 8]
```diff
@@ -16,70 +16,84 @@ import wandou.avpath.Parser.Syntax
...
     case object Insert extends Op
+
     case object InsertAll extends Op
+
   }
+
   private sealed trait Op
 
-  final case class Ctx(value: Any, schema: Schema, topLevelField: Schema.Field, target: Option[Target] = None)
+  final case class Ctx(value: Any, name: String, schema: Schema, topLevelField: Schema.Field, path: String, target: Option[Target] = None)
```
Comment (RyanSkraba) [inline src/main/scala/wandou/avpath/Evaluator.scala:65]: :/  Can you  comment (or even :gasp: javadoc) what name and path are?

I know: you'd be the first on in avpath to do it!

`path` is the original query, right?  and `name` is the *current* field name (if any) being processed?

File: src/main/scala/wandou/avpath/Evaluator.scala (hunk 4 above)
Comment (RyanSkraba) [reply src/main/scala/wandou/avpath/Evaluator.scala]:
```suggestion
            // We are on a Map, we name each sub element like the parent.  It should not be troublesome, but it's not right.
```

File: src/main/scala/wandou/avpath/Evaluator.scala (hunk 5 above)
Comment (RyanSkraba) [reply src/main/scala/wandou/avpath/Evaluator.scala]: :point_up: say it in the comment!

**Review by RyanSkraba — APPROVED** [review]

**Pull request merged**